### PR TITLE
Allow agit -o force-push work like -o force-push=true

### LINKED ISSUE #16679: agit flow hook refuses forced pushes
- Gitea version (or commit ref): `Gitea version 1.16.0+dev-96-gf1a810e09 built with GNU Make 4.3, go1.16.7 : netgo, osusergo, bindata, sqlite, sqlite_unlock_notify`
- Git version: `2.33.0.rc0.290.g2c90b51f76`
- Operating system: `Arch Linux`
- Built from: `source`
- Run with: `systemd`
- Database:
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] I have not tried yet, I don't see why it wouldn't work in a fresh repo like I tried it with...

- Log gist:
  - [ ] TBD

## Description
First off, I appreciate this feature (implemented recently in #14295), it is really handy (I didn't know about agit flow before btw).

I have run into an issue when force pushing an update into a PR while doing this agit flow stuff (after initially opening the PR using this method).
The force push was then refused by Gitea (a git hook).
The command used was `git push --force HEAD:refs/for/whatever-the-branch-was -o topic=topic-name` if I recall correctly.
EDIT: this was a typo here, I meant to say I pushed the `HEAD` ref, not "`HEADS`".

Note that
* branch protection has not been enabled in this particular instance,
* repo hooks have been reinitialized as per the subject PR request (it was noted somewhere there).

I cannot tell whether that is intended behaviour or not, i.e. whether I even ought to try to force push in agit flow or not.

### PR DESCRIPTION
**Pull request by lunny**
Fix #16679 

Docs update: https://gitea.com/gitea/docs/pulls/80

### CODE REVIEW

File: services/agit/agit.go [hunk 1]
```diff
@@ -25,7 +25,13 @@ import (
 func ProcReceive(ctx context.Context, repo *repo_model.Repository, gitRepo *git.Repository, opts *private.HookOptions) ([]private.HookProcReceiveRefResult, error) {
 	results := make([]private.HookProcReceiveRefResult, 0, len(opts.OldCommitIDs))
 	topicBranch := opts.GitPushOptions["topic"]
-	forcePush, _ := strconv.ParseBool(opts.GitPushOptions["force-push"])
+	forcePushStr := opts.GitPushOptions["force-push"]
```
Comment (6543) [inline services/agit/agit.go]:
```suggestion
	forcePushStr, forcePushOK := opts.GitPushOptions["force-push"]
```

File: services/agit/agit.go [hunk 2]
```diff
@@ -25,7 +25,13 @@ import (
...
+	var forcePush bool
+	if forcePushStr == "" {
```
Comment (6543) [inline services/agit/agit.go]:
```suggestion
	if forcePushOK && forcePushStr == "" {
```

Comment (6543) [reply services/agit/agit.go]: else force push will just be default ...

**Review by 6543 — APPROVED** [review]

**Review by a1012112796 — APPROVED** [review]

**Comment by wxiaoguang** [comment]
The code should be re-used, and other options should also use the better behavior.

And new code (at least `GitPushOptions.Bool`) should be tested.

image

**Comment by lunny** [comment]
> The code should be re-used, and other options should also use the better behavior.
> 
> And new code (at least `GitPushOptions.Bool`) should be tested.
> 
> image

I don't think we can use `GitPushOptions.Bool` here because `-o force-push` will get `false` but we think it should be `true`.

**Comment by wxiaoguang** [comment]
> > The code should be re-used, and other options should also use the better behavior.
> > And new code (at least `GitPushOptions.Bool`) should be tested.
>
> I don't think we can use `GitPushOptions.Bool` here because `-o force-push` will get `false` but we think it should be `true`.

The code should be re-used, and **other options should also use the better behavior**.

`GitPushOptions.Bool` should be improved but not introducing more unmaintainable code.

**Comment by wxiaoguang** [comment]
And one more question, how does `-o force-push` work? In code, I could only find `if len(kv) == 2 { hookOptions.GitPushOptions[kv[0]] = kv[1] }`

So, how could `force-push` be put into `GitPushOptions`?

**Comment by wxiaoguang** [comment]
> I don't think we can use `GitPushOptions.Bool` here because `-o force-push` will get `false` but we think it should be `true`.

And one more question, if you make  `-o force-push` work, then what `-o repo.private` should mean? Should it mean "repo.private" option exists but its value is false and the repo should be made as public?

That's a messy design.

**Comment by wxiaoguang** [comment]
I will write a correct one.

**Comment by wxiaoguang** [comment]
->  Make git push options accept short name #32245

**Pull request closed without merging**